File: src/parser.rs
use crate::types::{Node, NodeType, SessionEvent};
use anyhow::Result;

pub fn parse_event_to_node(event: SessionEvent) -> Result<Vec<Node>> {
    let mut nodes = Vec::new();
    let branch_level = if event.is_sidechain.unwrap_or(false) { 1 } else { 0 };

    if let Some(message) = event.message {
        match message.role.as_str() {
            "user" => {
                let text = extract_text_content(&message.content);

                // Check if it's a tool result
                if let Some(tool_results) = extract_tool_results(&message.content) {
                    for (idx, (tool_use_id, output, is_error)) in tool_results.iter().enumerate() {
                        nodes.push(Node {
                            id: format!("{}:{}", event.uuid, idx),
                            parent_id: Some(tool_use_id.clone()),
                            node_type: NodeType::ToolResult {
                                output: output.clone(),
                                is_error: *is_error,
                            },
                            timestamp: event.timestamp,
                            branch_level,
                            agent_id: event.agent_id.clone(),
                        });
                    }
                } else if !text.is_empty() {
                    nodes.push(Node {
                        id: event.uuid.clone(),
                        parent_id: event.parent_uuid.clone(),
                        node_type: NodeType::UserMessage(text),
                        timestamp: event.timestamp,
                        branch_level,
                        agent_id: event.agent_id.clone(),
                    });
                }
            }
            "assistant" => {
                let text = extract_text_content(&message.content);

                // Create assistant message node if there's text
                if !text.is_empty() {
                    nodes.push(Node {
                        id: event.uuid.clone(),
                        parent_id: event.parent_uuid.clone(),
                        node_type: NodeType::AssistantMessage(text),
                        timestamp: event.timestamp,
                        branch_level,
                        agent_id: event.agent_id.clone(),
                    });
                }

                // Extract tool uses
                if let Some(tool_uses) = extract_tool_uses(&message.content) {
                    for (_idx, (tool_id, tool_name, tool_input)) in tool_uses.iter().enumerate() {
                        nodes.push(Node {
                            id: tool_id.clone(),
                            parent_id: Some(event.uuid.clone()),
                            node_type: NodeType::ToolUse {
                                name: tool_name.clone(),
                                input: tool_input.clone(),
                            },
                            timestamp: event.timestamp,
                            branch_level,
                            agent_id: event.agent_id.clone(),
                        });
                    }
                }
            }
            _ => {}
        }
    }

    // Handle progress events
    if event.event_type == "progress" {
        nodes.push(Node {
            id: event.uuid.clone(),
            parent_id: event.parent_uuid.clone(),
            node_type: NodeType::Progress("Progress update".to_string()),
            timestamp: event.timestamp,
            branch_level,
            agent_id: event.agent_id.clone(),
        });
    }

    Ok(nodes)
}

fn extract_text_content(content: &serde_json::Value) -> String {
    match content {
        serde_json::Value::String(s) => s.clone(),
        serde_json::Value::Array(arr) => {
            arr.iter()
                .filter_map(|item| {
                    if let Some(obj) = item.as_object() {
                        if obj.get("type")?.as_str()? == "text" {
                            return obj.get("text")?.as_str().map(String::from);
                        }
                    }
                    None
                })
                .collect::<Vec<_>>()
                .join(" ")
        }
        _ => String::new(),
    }
}

fn extract_tool_uses(content: &serde_json::Value) -> Option<Vec<(String, String, String)>> {
    let arr = content.as_array()?;
    let mut tools = Vec::new();

    for item in arr {
        let obj = item.as_object()?;
        if obj.get("type")?.as_str()? == "tool_use" {
            let id = obj.get("id")?.as_str()?.to_string();
            let name = obj.get("name")?.as_str()?.to_string();
            let input = serde_json::to_string_pretty(obj.get("input")?).ok()?;
            tools.push((id, name, input));
        }
    }

    if tools.is_empty() { None } else { Some(tools) }
}

fn extract_tool_results(content: &serde_json::Value) -> Option<Vec<(String, String, bool)>> {
    let arr = content.as_array()?;
    let mut results = Vec::new();

    for item in arr {
        let obj = item.as_object()?;
        if obj.get("type")?.as_str()? == "tool_result" {
            let tool_use_id = obj.get("tool_use_id")?.as_str()?.to_string();
            let content = obj.get("content")?.as_str().unwrap_or("").to_string();
            let is_error = obj.get("is_error").and_then(|v| v.as_bool()).unwrap_or(false);
            results.push((tool_use_id, content, is_error));
        }
    }

    if results.is_empty() { None } else { Some(results) }
}

File: src/sources/claude/parser.ts
import type { SessionEvent, Node, NodeType } from "../../core/types"

let counter = 0
function generateId(): string {
  return `generated-${counter++}`
}

export function parseEventToNodes(event: SessionEvent): Node[] {
  const nodes: Node[] = []
  const branchLevel = 0
  const ts = new Date(event.timestamp).getTime()
  const uuid = event.uuid || generateId()

  if (event.message) {
    const { role, content } = event.message

    if (role === "user") {
      const toolResults = extractToolResults(content)
      if (toolResults) {
        for (let i = 0; i < toolResults.length; i++) {
          const [toolUseId, output, isError] = toolResults[i]
          nodes.push({
            id: `${uuid}:${i}`,
            parentId: toolUseId,
            nodeType: { kind: "tool_result", output, isError },
            timestamp: ts,
            branchLevel,
            agentId: event.agentId,
          })
        }
      } else {
        const text = extractTextContent(content)
        if (text) {
          nodes.push({
            id: uuid,
            parentId: event.parentUuid,
            nodeType: { kind: "user", text },
            timestamp: ts,
            branchLevel,
            agentId: event.agentId,
          })
        }
      }
    }

    if (role === "assistant") {
      const text = extractTextContent(content)
      const hasTextNode = text.length > 0
      const model = event.message?.model
      const usage = event.message?.usage

      if (hasTextNode) {
        nodes.push({
          id: uuid,
          parentId: event.parentUuid,
          nodeType: { kind: "assistant", text },
          timestamp: ts,
          branchLevel,
          agentId: event.agentId,
          model,
          usage,
        })
      }

      const toolUses = extractToolUses(content)
      if (toolUses) {
        const toolParent = hasTextNode ? uuid : event.parentUuid
        for (const [toolId, name, input] of toolUses) {
          nodes.push({
            id: toolId,
            parentId: toolParent,
            nodeType: { kind: "tool_use", name, input },
            timestamp: ts,
            branchLevel,
            agentId: event.agentId,
            model,
            usage,
          })
        }
      }
    }
  }

  // Skip progress events — they're streaming noise, not meaningful turns

  return nodes
}

function extractTextContent(content: unknown): string {
  if (typeof content === "string") return content
  if (!Array.isArray(content)) return ""
  return content
    .filter((item: any) => item?.type === "text" && typeof item?.text === "string")
    .map((item: any) => item.text as string)
    .join(" ")
}

// Returns [toolId, name, prettyInput][]
function extractToolUses(content: unknown): [string, string, string][] | null {
  if (!Array.isArray(content)) return null
  const tools: [string, string, string][] = []
  for (const item of content) {
    if (item?.type === "tool_use") {
      const id = item.id as string
      const name = item.name as string
      const input = JSON.stringify(item.input, null, 2)
      tools.push([id, name, input])
    }
  }
  return tools.length > 0 ? tools : null
}

// Returns [toolUseId, output, isError][]
function extractToolResults(content: unknown): [string, string, boolean][] | null {
  if (!Array.isArray(content)) return null
  const results: [string, string, boolean][] = []
  for (const item of content) {
    if (item?.type === "tool_result") {
      const toolUseId = item.tool_use_id as string
      const output = typeof item.content === "string" ? item.content : ""
      const isError = item.is_error === true
      results.push([toolUseId, output, isError])
    }
  }
  return results.length > 0 ? results : null
}
